doc/cascadia/Cascading-Default-Settings.md

```diff
@@ -0,0 +1,399 @@
+---
+author: Mike Griese @zadjii-msft
+created on: 2019-05-31
+last updated: 2019-06-13
+issue id: #754
+---
+
+# Cascading Default + User Settings
+
+## Abstract
+
+This spec outlines adding support for a cascading settings model. In this model,
+there are two settings files, instead of one.
+
+1. The default settings file
+2. The user's settings file
+
+The default settings file would be a static, read-only file shipped with the
+terminal. The user settings file would then contain all the user's chosen
+customizations to the settings. These two files would then be composed together
+when the app is launched, so that the runtime settings are the union of both the
+defaults and whatever modifications the user has chosen. This will enable the
+app to always use a default schema that it knows will be valid, and minimize the
+settings that the user needs to customize.
+
+Should the settings schema ever change, the defaults file will change, without
+needing to re-write the user's settings file.
+
+
+## Inspiration
+
+Largely inspired by the settings model that both VS Code (and Sublime Text) use.
+
+### Goal: Minimize Writing to `profiles.json`
+
+We want to write the user settings file, `profiles.json`, as little as possible.
+Each time we write the file, there's the possiblity that we've re-ordered the
+keys, as `jsoncpp` provides no ordering guarantee of the keys. This isn't great,
+as each write of the file will randomly re-order the file.
+
+One of our overarching goals with this change should be to re-write the user
+settings file as little as possible.
+
+## Solution Design
+
+The settings are now composed from two files: a "Default" settings file, and a

```

Comment:
Should we call out dynamic here and in the abstract above?

Reply:
eh - I added it to the abstract, but I think it doesn't totally belong here. Dynamic profiles are just a special type of profile in the user settings.

Reply:
My mental model tells me dynamic profiles are just a special type of profile in the _default_ settings. It's like having 100% of cmd's profile in `default.json` but it's fully generated at runtime/not from a json file

Reply:
Oh see, I made them something that's generated at runtime, _into the `profiles.json`_ file. That way, they can be created dynamically, but once they're created, they can be modified, and those modifications will persist to future launches.

Reply:
I am now confused.

For this example WSL profile, what ends up in the `profiles.json`?

### whole profile

```json
        {
            "guid" : "{1caa0dad-35be-5f56-a8ff-afceeeaa6101}", // WSL Ubuntu
            "acrylicOpacity" : 0.75,
            "closeOnExit" : true,
            "colorScheme" : "Solarized Dark",
            "commandline" : "wsl.exe -d Ubuntu",
            "cursorColor" : "#FFFFFF",
            "cursorShape" : "emptyBox",
            "fontFace" : "Consolas",
            "fontSize" : 10,
            "historySize" : 9001,
            "icon" : "ms-appx:///ProfileIcons/{9acb9455-ca41-5af7-950f-6bca1bc9722f}.png",
            "name" : "WSL",
            "padding" : "0, 0, 0, 0",
            "snapOnInput" : true,
            "startingDirectory" : "%USERPROFILE%",
            "useAcrylic" : true
        }
```

### overrides only

```json
        {
            "guid" : "{1caa0dad-35be-5f56-a8ff-afceeeaa6101}", // WSL Ubuntu
            "cursorColor" : "#FF0000",
            "cursorShape" : "filledBox",
        }
```

Reply:
I'm thinking when the profile is generated, only the following is added to the `profiles.json` file:
```json
        {
            "guid" : "{1caa0dad-35be-5f56-a8ff-afceeeaa6101}",
            "icon" : "ms-appx:///ProfileIcons/{9acb9455-ca41-5af7-950f-6bca1bc9722f}.png",
            "commandline" : "wsl.exe -d Ubuntu",
            "name" : "Ubuntu",
        }
```
These are the properties that are different from the defaults, though now that I'm thinking about it, I should maybe call out something about what the profile defaults are. Because I want the above to be all that's in the file (at the start), but I've made the assumption that this will cleanly be separate from what values are in a default-constructed `Profile`. 

I call out only serializing things that are different from things in the `defaults.json`, but not really the fact that I don't think we _need to_ serialize things that are the same as the default values for that object type.

However, as you've called out (somewhere else), we don't really have a _default profile_ object that the user can change to change all unset profile settings. I think that might be out of scope for this changelist, yet at the same time is something that should probably be considered as a part of this. I'll mull on that.

Reply:
I think that lends more towards thinking about the dynamic as a type of default, rather than a type of user-profile. So, the dynamic provider wuold just _make all its dynamics_ and then the user profile laoder would override bits and pieces, and hide and show them.

This is my mental model:
![Cascading](https://user-images.githubusercontent.com/14316954/59538392-6ca95080-8eae-11e9-9f95-2d744585b54d.png)


I don't think we actually _have_ cascading settings without the blue text (user default), and if user profiles.json _needs to_ contain any of the vital parts of the dynamic. I dunno, more noodling is required. 😄

Reply:
Okay, spoilers from the next revision:
1. Load the `defaults.json` (the default settings)
1. Load the `profiles.json` (the user settings)
1. Perform a preliminary scan of the user settings, and create all the profiles
   in order they appear in the user settings file.
1. Layer all settings from the defaults on the existing profiles, and create the
   default color schemes, as well as the default global settings.
1. Layer all user settings upon the existing settings models.
1. Generate any dynamically generated profiles, if necessary. This step might
   request the user settings are saved out.
1. Validate the settings. This step might request the user settings are saved
   out.
1. If necessary, write the modified settings back to `profiles.json`.

Reply:
(previously generated dynamic profiles are included in step 5, as if they're just another user profile)

Reply:
I still don't understand why 6 can't happen after 1 before 2. ... if we pretend mentally that a dynamic profile _was just another blob of defaults.json content_, it makes everything line up. Let's hash it out on Monday :)

Reply:
Oh, I think there's a miscommunication there - by "load the X settings", I mean, "read them from disk to a `JsonValue` object".

We need to be able to query the composed settings (after 5) to know which dynamic generators we should run.

Honestly, this entire debate seems like it's arguing over whether the dynamic profile generator should run after the defaults are applied (step 4) or if it should run after the user profiles are layered (step 5)... I don't really think there's a ton of difference to those two options (save for the whole "we need the user's `autoload*Profiles` setting" issue).

Reply:
With #1321 merged in, I think this is resolved now